File: chapt5/chapt5_ws/src/demo_python_tf/demo_python_tf/dynamic_tf_broadcaster.py
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import TransformStamped
from tf2_ros import TransformBroadcaster
from tf_transformations import quaternion_from_euler


class DynamicTFBroadcaster(Node):
    def __init__(self):
        super().__init__('dynamic_tf2_broadcaster')
        self.dynamic_broadcaster = TransformBroadcaster(self)
        # self.pub_dynamic_tf()
        self.timer = self.create_timer(0.1, self.pub_dynamic_tf)

    def pub_dynamic_tf(self):
        # Create a TransformStamped message
        t = TransformStamped()

        # Set the header frame_id and child_frame_id
        t.header.stamp = self.get_clock().now().to_msg()
        t.header.frame_id = 'camera_link'
        t.child_frame_id = 'bottle_link'

        # Set the translation (x, y, z)
        t.transform.translation.x = 0.2
        t.transform.translation.y = 0.0
        t.transform.translation.z = 0.5

        # Set the rotation (roll, pitch, yaw)
        roll = 0.0
        pitch = 0.0
        yaw = 0.0
        quaternion = quaternion_from_euler(roll, pitch, yaw)
        
        t.transform.rotation.x = quaternion[0]
        t.transform.rotation.y = quaternion[1]
        t.transform.rotation.z = quaternion[2]
        t.transform.rotation.w = quaternion[3]

        # Publish the dynamic transform
        self.dynamic_broadcaster.sendTransform(t)
        self.get_logger().info('Published dynamic transform from world to robot_base')
    
def main(args=None):
    rclpy.init(args=args)
    node = DynamicTFBroadcaster()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

File: chapt5/chapt5_ws/src/demo_cpp_tf/src/dynamic_tf_broadcaster.cpp
#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "tf2/LinearMath/Quaternion.h"
#include "tf2_ros/transform_broadcaster.h"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
#include <chrono>

using namespace std::chrono_literals;


class DynamicTfBroadcasterNode : public rclcpp::Node
{
public:
  DynamicTfBroadcasterNode()
  : Node("dynamic_tf_broadcaster")
  {
    dynamic_broadcaster_ = std::make_shared<tf2_ros::TransformBroadcaster>(this);
    timer_ = this->create_wall_timer(
      10ms, std::bind(&DynamicTfBroadcasterNode::publish_tf, this));
  }

  void publish_tf()
  {
    geometry_msgs::msg::TransformStamped dynamic_transform_stamped;

    dynamic_transform_stamped.header.stamp = this->get_clock()->now();
    dynamic_transform_stamped.header.frame_id = "map";
    dynamic_transform_stamped.child_frame_id = "base_link";

    dynamic_transform_stamped.transform.translation.x = 2.0;
    dynamic_transform_stamped.transform.translation.y = 3.0;
    dynamic_transform_stamped.transform.translation.z = 0.0;

    tf2::Quaternion quat;
    quat.setRPY(0, 0, 30 * M_PI / 180);  // Roll, Pitch, Yaw
    dynamic_transform_stamped.transform.rotation = tf2::toMsg(quat);

    dynamic_broadcaster_->sendTransform(dynamic_transform_stamped);
    RCLCPP_INFO(this->get_logger(), "Published dynamic transform from 'map' to 'base_link'");
  }
private:
  std::shared_ptr<tf2_ros::TransformBroadcaster> dynamic_broadcaster_;      
  rclcpp::TimerBase::SharedPtr timer_;
};  

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<DynamicTfBroadcasterNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}






